### ISSUE TITLE
deploy.rb template should use pessimistic version operator for `lock`

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1930]
Assuming I am running Capistrano 3.9.1, then `cap install` generates the following in `deploy.rb`:

```ruby
lock '3.9.1'
```

Relevant code: https://github.com/capistrano/capistrano/blob/bf45cbafe5ef396933c81f334bdbc8c6d32a38c2/lib/capistrano/templates/deploy.rb.erb#L2

This means that my project can *only* be deployed using version 3.9.1.

This seems unnecessarily strict. If the Capistrano maintainers are doing a good job of adhering to the spirit of semver, then it should also be safe to deploy with 3.9.2, 3.9.3, etc. Arguably, I should be able to use *any* 3.x version, like 3.10, 3.11, etc. (Although Capistrano is only loosely following semver, so this might be a bit risky. See #1894.)

I can achieve this by changing my `deploy.rb` as follows:

```ruby
# Patch releases are OK
lock '~> 3.9.1'
```

Or more broadly:

```ruby
# Feature bumps are OK
lock '~> 3.9'
```

(The `~>` operator is called the "pessimistic version" operator in rubygems and Bundler.)

I would like to make the change to the `deploy.rb` template to use `~>`. The question is, should we lock to the patch level (`x.y.z`) or the minor (`x.y`) level?

**Comment by will-in-wi** [comment]
The tricky part is that I would say minor with bundler and patch without…

**Comment by will-in-wi** [comment]
At a bare minimum, I'd suggest we at least do patch for the next release. This is a pain point, methinks. If we want to expand it to minor, that'd be fine.

**Comment by will-in-wi** [comment]
We could do patch, but have the comment suggest dropping it to minor if you use bundler… Or possibly link to the to-be-written-docs: https://github.com/capistrano/capistrano/issues/1929

**Comment by jrochkind** [comment]
If there are really no backwards incompatibilities between major versions, major/minor should be sufficient? `"~> 3.2"` means "at least 3.2.0, but less than 4.0"

**Comment by mattbrictson** [comment]
> If there are really no backwards incompatibilities between major versions

@jrochkind in the PR I linked above, this clarification was added to the docs:

> For a `X.Y.Z` release... `Y` is bumped when we add new features. **Occasionally a `Y` release may include small breaking changes**

(emphasis added)

**Comment by jrochkind** [comment]
I'd strongly suggest considering, well, not doing that. But otherwise, if they are as occasional and 'small' as you intend, it's perhaps still fine to, at least as a default generated thing, lock only major version.  For the same reasons y'all think it's okay to occasionally include such small breaking changes, right?

Is there a reason I'm missing it makes a difference if you are using bundler or not, as to how strict a lock is advisable?

**Comment by will-in-wi** [comment]
In my use of Bundler, I'm going to have to run `bundle update` to upgrade Capistrano. So I'm okay with Capistrano being really loose on its enforcement, because if I ran into an issue, I'd fix it in the Gemfile.

If I didn't have Bundler (a case I have not yet encountered; I use Bundler for managing Capistrano in my non-Ruby app deployments), I would be more worried about a Bundler managed app magically installing a much newer version of Capistrano and breaking my deployment next time I used it.

Perhaps the issue is exactly the semver one. We might just want to lock the major version and do our best not to break things. I'm personally of the opinion that it's not a bad idea to increment the major version more frequently, once or twice a year in an app with this level of churn. But Capistrano 2 -> 3 was a big deal (I have no idea about 1 -> 2), and people might be worried about moving from 3 -> 4, or expect it to be a big change.
